doc/cpp_recipe.rst

```diff
@@ -0,0 +1,230 @@
+    * The model description (i.e. recipe) should be seperate from the
+      instantiation of the model to execute on the target system.
+    * To minimise the amount of memory used in model building, to make it
+      possible to build and run simulations in one run.
+
+Recipe descriptions are cell-oriented, in order that the building phase can
+distributed and that the recipe description can be built indepedently of any
+runtime execution environment.
+
+During model building, the recipe is queried first by a load balancer,
+then later when building the low-level cell groups and communication network.
+The cell-centered recipe interface, whereby cell and network properties are
+specified "per-cell", facilitates this.
+
+The steps of building a simulation from a recipe are:
+
+.. topic:: 1. Load balancing
+
+    First, the cell `gid` are partitioned cells over MPI ranks, and each rank
+    the cells to build up a cost model for the cells it has been assigned 
+    Then, the ranks coordinate to redistribute cells over MPI ranks so that
+    each rank has a balanced workload. Finally, each rank groups its local
+    cells into :cpp:type:`cell_group` s that balance the work over threads (and
+    GPU accelerators if available).
+
+.. topic:: 2. Model building
+
+    The model building phase takes the cells assigned to the local rank, and builds the
+    local cell groups and the part of the communication network by querying the recipe
+    for more information about the cells assigned to it.
+
+.. _recipe_best_practice:
+
+Best Practices
+--------------
+
+Here is a set of rules of tumb to keep in mind when making recipes. The first is
+mandatory, and following the others as closely as possible will lead to better
+performance.
+
+.. topic:: Stay thread safe
+
+    The load balancing and model construction are multithreaded, i.e.
+    multiple threads query the recipe simultaneously.
+    So calls to a recipe member should not have side effects, and use
+    lazy evaluation when possible (see `Be lazy <recipe_lazy_>`_).
+
+.. _recipe_lazy:
+
+.. topic:: Be lazy
+
+    A recipe does not have to contain a complete description of the model in
+    memory, instead it should precompute as little as possible, and use
+    `lazy evaluation <https://en.wikipedia.org/wiki/Lazy_evaluation>`_ to generate
+    information only when requested.
+    This has multiple benefites, including:
+
+        * thread safety;
+        * minimise memory footprint of recipe.
+
+.. topic:: Think of the cells
+
+    When formulating a model, think cell-first, and try to formulate the model and
+    the associated workflow from a cell-centered perspective. If this isn't possible,
+    please contact the developers, because we would like to develop tools that help
+    make this simpler.
+
+.. topic:: Be reproducable
+
+    Arbor is designed to give reproduceable results when the same model is run on
+    different number of MPI ranks, or threads, or hardware (e.g. GPUs).
```

Comment:
* _on a different number of_ or _on different numbers of_
* _different hardware_ (otherwise it reads as _different number of … hardware_).

Reply:
Still reads oddly.